### PR TITLE
[SPARK-11063][Streaming]Change preferredLocations of Receiver's RDD to hosts rather than hostports

### PR DESCRIPTION
**Pull request by zsxwing**
The format of RDD's preferredLocations must be hostname but the format of Streaming Receiver's scheduling executors is hostport. So it doesn't work.

This PR converts `schedulerExecutors` to `hosts` before creating Receiver's RDD.

### CODE REVIEW

**Comment by jerryshao** [comment]
The fix looks good to me.

**Comment by tdas** [comment]
What is the implication of this bug in case for the scheduling? We are trying to evenly distribute receivers to executors. But if we set the preferredLocations to only the granularity of hosts, then the following could happen. Two executors can be in the same host (through YARN). We set the preferred locations for two receivers for that same host even though we want them to be in two different executors. It may so happen that both executors get scheduled at the same executor. Isnt it?

**Comment by zsxwing** [comment]
> What is the implication of this bug in case for the scheduling? We are trying to evenly distribute receivers to executors. But if we set the preferredLocations to only the granularity of hosts, then the following could happen. Two executors can be in the same host (through YARN). We set the preferred locations for two receivers for that same host even though we want them to be in two different executors. It may so happen that both executors get scheduled at the same executor. Isnt it?

Right. However, `preferredLocations` only supports hosts. So at least for branch 1.5, this is the best can can do. We cam enhance `preferredLocations` to support host and port in master branch.

**Comment by jerryshao** [comment]
From my understanding, Spark's scheduler does not support scheduling with port (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/TaskLocation.scala)

**Comment by tdas** [comment]
Yeah, we have to fix it for branch 1.5 with a small fix like this. 
Is it possible to test this with unit tests? Could you check if there are unit tests of the DAGScheduler / TaskScheduler that we can use to unit test so that something like this does not happen ever again.

**Comment by jerryshao** [comment]
I think we could add another `TaskLocation`, like `ExecutorTaskLocation` and pass the `preferredLocations` with special tag prefix like HDFS in memory tag, to let TaskManager now we want to schedule the task with process local.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 1]
```diff
@@ -388,7 +388,7 @@ private[spark] class TaskSetManager(
     if (TaskLocality.isAllowed(maxLocality, TaskLocality.NO_PREF)) {
       // Look for noPref tasks after NODE_LOCAL for minimize cross-rack traffic
       for (index <- dequeueTaskFromList(execId, pendingTasksWithNoPrefs)) {
-        return Some((index, TaskLocality.PROCESS_LOCAL, false))
+        return Some((index, TaskLocality.NO_PREF, false))
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: @CodingCat could you take a look at this change?

**Comment by zsxwing** [comment]
Added a unit test and fixed a minor issue in `TaskSetManager`.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 1 above)
Comment (CodingCat) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: sure, will do tomorrow morning

**Comment by zsxwing** [comment]
Tested this patch in a 5 nodes cluster. Each node has an executor, the executor core is 1 and the receiver number is 5. 

Before this patch, there were several Receiver restarting logs. After applying this patch, all restarting logs disappeared.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 1 above)
Comment (tdas) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: This looks like a significant bug in the core. If so probably go into a separate patch.

Comment (tdas) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: And how is this related to the main bug we are considering in this PR?

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: This is not related. I will move it in a separate PR.

Comment (CodingCat) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: hmmm....it's a typo but would not take any effect 

see discussions here:  https://github.com/apache/spark/pull/3816 (only look at the comments after Dec 30, 2014 is enough)

Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Didn't notice this PR. Thanks.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala [hunk 2]
```diff
@@ -80,6 +82,28 @@ class ReceiverTrackerSuite extends TestSuiteBase {
...
+          receiverTaskLocality = taskStart.taskInfo.taskLocality
+        }
+      })
+      val input = ssc.receiverStream(new TestReceiver)
+      val output = new TestOutputStream(input)
+      output.register()
+      ssc.start()
+      eventually(timeout(10 seconds), interval(10 millis)) {
+        // If preferredLocations is set correctly, receiverTaskLocality should be NODE_LOCAL
+        assert(receiverTaskLocality === TaskLocality.NODE_LOCAL)
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: What will it be if it is not set correctly? That is, without the fix above, what would locality be

**Comment by tdas** [comment]
LGTM. Just a clarifying question.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala (hunk 2 above)
Comment (zsxwing) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/ReceiverTrackerSuite.scala]: It's `TaskLocality.ANY` without this patch.

**Comment by tdas** [comment]
Merging this to master and branch 1.5
